### ISSUE TITLE
Prevent shadowing globals

### ISSUE THREAD

**Issue by kasvtv** [eslint/eslint#13019]
**Please describe what the rule should do:**
I've shot myself in the foot more times than I dare to admit by creating a variable called `process` for example, then wondering why further up in my code `process.env.API_KEY` or something wasn't working. Clearly, it's not a smart idea to shadow any global variables, I would love a rule for this that forbids one to declare variables or function params that have the same name as a global. Which globals there are should be able to be inferred from the `env` portion of the config.

I've tried a lot of things to see if I could get this behaviour from ESlint, but couldn't find a way to do it, please correct me if I'm wrong.
**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// incorrect code
const console = { asdf: 123 }
const JSON = { asdf: 123 }
const process = { asdf: 123 }
const exports = { asdf: 123 }
```

**Why should this rule be included in ESLint (instead of a plugin)?**
I think a rule like this would be obviously useful to most people as it's a pretty clear and common mistake.

**Are you willing to submit a pull request to implement this rule?**
I unfortunately don't have enough time to familiarize myself with the codebase

**Comment by yeonjuan** [comment]
Hi, @kasvtv 
> I've tried a lot of things to see if I could get this behaviour from ESlint, but couldn't find a way to do it, please correct me if I'm wrong.

Maybe you can find the way by using the rule [no-shadow](https://eslint.org/docs/rules/no-shadow) with the option `{ "builtinGlobals": true }`.
* [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2hhZG93OiBbXCJlcnJvclwiLCB7IFwiYnVpbHRpbkdsb2JhbHNcIjogdHJ1ZSB9XSovXG4vKmVzbGludC1lbnYgbm9kZSwgYnJvd3NlciAqL1xuY29uc3QgY29uc29sZSA9IHsgYXNkZjogMTIzIH1cbmNvbnN0IEpTT04gPSB7IGFzZGY6IDEyMyB9XG5jb25zdCBwcm9jZXNzID0geyBhc2RmOiAxMjMgfVxuY29uc3QgZXhwb3J0cyA9IHsgYXNkZjogMTIzIH0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7ImJyb3dzZXIiOmZhbHNlLCJub2RlIjpmYWxzZX19fQ==)
```js
/*eslint no-shadow: ["error", { "builtinGlobals": true }]*/

const console = { asdf: 123 } // Error - 'console' is already declared ...
const JSON = { asdf: 123 }    // Error - 'JSON' is already declared ...
const process = { asdf: 123 } // Error - 'process' is already declared ...
const exports = { asdf: 123 } // Error - 'exports' is already declared .

```

**Comment by kasvtv** [comment]
Thanks so much! Hopefully other people wanting the same thing will stumble upon this issue.

**Comment by mdjermanovic** [comment]
Closing since the question has been answered.
